### DISCUSSION TITLE
Trying to calculate "% played" for video tracks but get type error "has type float?"

### DISCUSSION THREAD

**Discussion by Moonbase59** [savonet/liquidsoap#3149] [Q&A] [answered]
Again I’m trying to do something simple and sat half the night, to no avail: I want to calculate (and later show) the "percent played" of the current video track.

This is the code so far:
```
# Get video duration on new track
v_duration = ref(0.0)
def update_duration(m)
  v_duration := request.duration(m['filename'])
end

def v_percent(s)
  print("Elapsed: #{source.elapsed(s)} of #{v_duration()}")
  if v_duration() > 0.0 then
    # Error 5: this value has type float? (inferred at videostream4.liq, line 69, char 4-21) but it should be a subtype of something that is a number type
    source.elapsed(s) / v_duration() * 100.0
  else
    0.0
  end
end

#s = video.testsrc(width=1280, height=720)
s = mksafe(playlist(reload_mode="watch", video_playlist))
s.on_track(update_duration)
```

What am I doing wrong?

It seems `source.elapsed(s) / v_duration()` is of type `float?` and should be something else (`float`, I suppose). When commenting out the division, the `print` shows a nicely incrementing float, and the duration is also correct after about 0.75s (null before).

**Comment by smimram** [comment, marked as answer]
`request.duration` can return `null` if the computation of the duration failed. You should do `null.get(requestion.duration(...))` to get the value.

**Comment by Moonbase59** [comment]
Thanks, that brought me further! is there a faster/better way to get a video track’s duration?

Next question, if I may… The goal of all that "% played" calculation is to show an unobtrusive small bar below the text box, to let the listener know how far we have progressed in the currently playing video:

I thought of using `video.add_rectangle()` for this:
```
# make a progress bar below the text (added 100 to the width to make it visible)
s = video.add_rectangle(color=0x80ebebeb, x=0, y=670, height=4, width=100+int(1280/100*v_percent(v)), s)
# for diagnosis only:
s = video.add_text(font=font_file, size=20, x=50, y=680, color=0xebebeb, {string(v_percent(v)) ^ "% of " ^ string(v_duration()) ^ " (" ^ v_title() ^ ")"}, s)
```
where source `v` is the video playlist, source `s` the muxed source with station audio and text added.

![video m2t - VLC media player_004](https://github.com/savonet/liquidsoap/assets/3706922/3ca9b56b-376c-4625-912c-bf3e47090619)

Unfortunately, the "rectangle", i.e. "progress bar" _never updates_. I wonder what I’m missing this time.

**Comment by Moonbase59** [comment]
Minimal example:

```
v = video.testsrc(width=1280, height=720)

# make a progress bar below the text
v = video.add_rectangle(color=0x8000eb00, x=0, y=670, height=4, width=random.int(min=0, max=1279), v)
# same with text
v = video.add_text(size=20, x=50, y=680, color=0x00eb00, {string(random.int(min=0, max=1279))}, v)

output.audio_video(v)
```

**Comment by Moonbase59** [comment]
Ha! Found it! For whatever reason, I have to use curly brackets around the part that should calculate new everytime:

```
v = video.add_rectangle(color=0x8000eb00, x=0, y=670, height=4, width={random.int(min=0, max=1279)}, v)
```

Now I have to find out what the curlies actually _mean_ in LS…

**Comment by vitoyucepi** [comment]
Hi @Moonbase59,
This piece of code looks incorrect to me.
```
# Get video duration on new track
v_duration = ref(0.0)
def update_duration(m)
  v_duration := request.duration(m['filename'])
end
```
I think most of the sources support `source.duration`.
Probably it's possible to use `s.duration`, where `s` is of type `source`.

**Comment by Moonbase59** [comment]
@vitoyucepi Hi, thanks for the suggestion! I was never really happy with having to read a file from metadata just to get the duration, but didn’t know better. ;-)

Using `source.duration()` I seem to get the same, and faster, plus it seems to work on a track basis, so I can throw away all that `on.track` and stuffing the duration into a "variable" stuff. So much more elegant, reduces the percentage calculation from

```
# Get video duration on new track
v_title = ref("")
v_duration = ref(0.0)
def update_duration(m)
  v_title := m["title"]
  # request.duration can return null if the computation of the duration failed
  v_duration := null.get(default=0.0, request.duration(m['filename']))
end

# calculate % played of current video track
def v_percent(s)
  # avoid v_duration being not ready yet, don't divide by zero
  if v_duration() > 0.0 then
    source.elapsed(s) / v_duration() * 100.0
  else
    0.0
  end
end
…

v.on_track(update_duration)
```
to just
```
# calculate % played of current video track
def v_percent(s)
  source.elapsed(s) / source.duration(s) * 100.0
end
```
Much better, and rather happy with the result:

![video m2t - VLC media player_005](https://github.com/savonet/liquidsoap/assets/3706922/f640d1e7-ea4a-4748-b454-c677eaa29eb3)
![video m2t - VLC media player_006](https://github.com/savonet/liquidsoap/assets/3706922/d665fa71-7f26-4bdd-8a84-7cf09cecf80f)

_Note to myself:_ The important thing here was to keep the sources separate: Source `a` for the original audio playlist (or radio stream), source `v` for the original video playlist (on which the "played %" gets shown), and source `s` which holds the video plus the muxed-in radio stream as well as the overlays. If using only _one_ source, the progress indicator whould show progress of the radio tracks playing (if known; doesn’t work for streams). Using `v_percent(a)` and `v_percent(v)` we could show audio & video progression independently.

**Reply by Mikulskii** [reply]
Hey, thank you so much for such a cool idea!
Today my friends and I tried to reproduce this feature and as a result of brainstorming we came to this single line:
```
s = video.add_rectangle(color=0xfcb900, x=0, y=715, height=10, width={int(1280.0*source.elapsed(s) / source.duration(s))}, s)
```
In my case, this is the only source with music videos.

**Reply by Moonbase59** [reply]
Hey @Mikulskii, what a coincidence… I accidentally ran across [your blog post](https://mikulski.rocks/creating-a-24-7-lofi-like-stream-part-2/) which in turn led me back here… :rofl: 

You’re of course right. With the ever-evolving Liquidsoap, things are often a little trial-and-error, and I admit I’m often thinking too complicated, trying to make evolving code do much more—and ending up not using it.

Yours is much simpler and does the job as well! Thanks for the feedback!

**Reply by Mikulskii** [reply]
Ahah, cool - the circle has closed :D 
Thank you!
